keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,273 +122,289 @@ consistent with the syntax supported by any of the common proxy vendors:
-from a v1beta1 where `path` was used, we set the annotation.
-Validation logic will check the annotation to decide how to interpret
-and validate paths.  We'll have to keep it for a long while.  Maybe we
-can prevent new objects from being created with that annotation?  Now
-what if implementations want to offer "extended" semantics here?  Same
-validation problem.
-
-3)  v1beta1 gets a "PathPrefix" field as above.  v1 gets an annotation
-"ingress.kubernetes.io/non-portable-path-semantics".  If we create a
-v1 object from a v1beta1 where `path` was used, we set the annotation.
-We keep it forever.
-
-4) Same as 2 but we just don't validate Path at all.
--->
-
-#### Rejected alternatives
-
-##### Portable regex behavior
-
-The safest route for specifying the regex would be to state a limited
-subset that can be used in a portable way. Any expressions outside of
-the subset will have implementation specific behavior.
-
-Regular expression subset (derived from [re2][re2-syntax] syntax page)
-
-| Expression | description             |
-|------------|-------------------------|
-| `.`        | any character           |
-| `[xyz]`    | character class         |
-| `[^xyz]`   | negated character class |
-| `x*`       | 0 or more x's           |
-| `x+`       | 1 or more x's           |
-| `xy`       | x followed by y         |
-| `x|y`      | x or y (prefer x)       |
-| `(abc)`    | grouping                |
+  // Defaults to "/" if empty.
+  //
+  // +Optional
+  Path string
 
-Maintaining a regular expression subset is likely not worth the complexity and
-is likely impossible across the [many implementations][regex-survey].
+  // PathType determines the interpretation of the Path
+  // matching. PathType can be one of the following values:
+  //
+  // Exact  - matches the URL path exactly.
+  // Prefix - matches based on a URL path prefix split
+  //   by '/'. [insert description of semantics described below]
+  //
+  // ImplementationSpecific - interpretation of the Path
+  //   matching is up to the IngressClass. Not all
+  //
+  // +Optional
+  PathType string
+  ...
+ }
+ ```
+
+Interoperability between versions will be as follows:
+
+1. v1beta1 will default to `ImplementationSpecific`.
+1. v1 will default to `Prefix`.
+
+Note: default value are permitted between API versions
+([reference][api-conv-versions]).
+
+[api-conv-versions]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#defaulting
+
+##### Path matching semantics
+
+For non-`ImplementationSpecific` paths:
+
+1. Let `[p_1, p_2, ..., p_n]` be the list of Paths for a specific host.
+1. Every Path `p_i` must be syntactically valid:
+    1. Must begin with the `'/'` character (relative paths are not allowed by [RFC-7230][rfc7230]).
+    1. Must not contain consecutive `'/'` characters (e.g. `/foo///`, `//`).
+1. A trailing `'/'` character in the Path is ignored, e.g. `/abc` and `/abc/`
```

Comment:
This is not true in all LBs, right? If this is part of our spec, will we force those LBs to carry 2 rules for every match?

Reply:
There needs to be some transformation at some level. The generated configurations are somewhat munged (see the Lua-based matches in Nginx), so the cost of two rules does not seem like an excessive cost to pay for consistency.